### DISCUSSION TITLE
Is there an up-to-date springboot version of LibreClinica?

### DISCUSSION THREAD

**Discussion by shanzhu2021** [reliatec-gmbh/LibreClinica#216] [General]
I'd like to know is there an up-to-date springboot version of LibreClinca which I can open and run in my IntelliJ IDEA.

**Comment by chae-reliatec** [comment]
Hi shanzhu2021

So your question about LibreClinica is whether you can not run it from within IntelliJ IDEA like you are used from Spring Boot projects or not, right?
Since the roots of LibreClinica are much older then Spring Boot it uses Spring in the "traditional" way. If you like to volunteer to migrate LibreClinica to Spring Boot we can start a discussion with the contributors whether they find it useful or not.

Regards,
Christian

**Comment by shanzhu2021** [comment]
Yes, please start this process, I 'd like to work for the migration of LibreClinica to Spring Boot.

**Comment by toskrip** [comment]
My question would be if you are interested about spring boot version only so that you can run it in development or if you want to run LibreClinica as spring boot app in production. Because if you are only searching for a way how to setup development environment with intellij (java, postgresql, maven build, run in debug mode), then this is of course possible without spring boot and I will update our developer documentation with tutorial how to do it.

**Comment by shanzhu2021** [comment]
My answer is both - first I need a clear way to setup development environment with intellij, please update the developer documentation with tutorial how to do it, including the SQL scripts on initializing the database tables, and I also want to run LibreClinica as spring boot app in production in the near future, with all the dependencies and plugins up-to-date.

**Comment by toskrip** [comment]
Ok. As I mentioned, I will be working on dev documentation so I can add some details regarding how to setup development environment. Just briefly, for developer it should be pretty straightforward. You import it as a maven project. Setup your local jdk and tomcat so that you can create your tomcat specific run/debug configuration. Use maven to execute specific goals. Once build, it produces war that you define in your build/debug configuration to be deployed in tomcat. Regarding database initialisation. Liquibase is used for DB scheme version management. The initialisation/update of database scheme is perform automatically on deployment (only empty DB is required at beginning).

best

Tomas

**Comment by toskrip** [comment]
I did a bit of work on dev documentation. Hopefully it will help you to get started:

[libreclinica-docs.readthedocs.io](https://libreclinica-docs.readthedocs.io)

best

Tomas

**Comment by shanzhu2021** [comment]
Dear Tomas,


I am trying to follow your dev documentation, but I use an Intellij IDEA community Edition, I cannot find a Server tab at the Run/Debug Configuration step. Is there other way to do it?


Regards,
Tong

**Comment by toskrip** [comment]
I am not sure if I would recommend IntelliJ IDEA Community Edition for this project, as far as I know it does not have support for Java EE and other frameworks used in enterprise application development. And it seem to also not have the Tomcat plugin necessary to setup Tomcat server as described in the documentation. Quick search on the internet says that it may be possible to configure Tomcat as external tool, but I cannot share my experience in this direction.

**Comment by shanzhu2021** [comment]
I switch to Ultimate edition, and open the project in IntelliJ IDEA, finally there are several errors in pom file as follows:
Line 450: <finalName>${assembly.name}-${project.version}</finalName>
Line 469: <home>${catalina.home}</home>
Line 472: <home>${catalina.home}</home>
and report:
Could not find artifact oracle.ojdbc14:ojdbc14:pom:10g in openclinica.jfrog.io-libs-release (https://openclinica.jfrog.io/openclinica/libs-release)
please help!
<img width="656" alt="errors" src="https://user-images.githubusercontent.com/77312250/105273180-beab6f80-5bd5-11eb-84b0-7f33e6cee9d1.PNG">

**Comment by toskrip** [comment]
the lines 450, 469, 472 belong to maven plugins that are not used if developer machine is setup as described in docs. The error only says that this maven plugin would not work, but these two plugins are not used (at least not in the parent project level). 

The Oracle JDBC driver is more of relic (Oracle DB backend is not supported in the system for long time already). We plan to remove it in one of the next refactoring branches, but currently there is still code (although unused) which operates with it. The dependency is only available from OC server (which is also one of the reasons why we want to get rid of it). If for some reason (network/proxy...) you (or maven) cannot reach that server then you have a problem. Can you reach this endpoint? 

https://openclinica.jfrog.io/artifactory/libs-release/oracle/ojdbc14/ojdbc14/10g/ojdbc14-10g.jar

**Comment by shanzhu2021** [comment]
Yes, I can get it but very slow.
So I should ignore all these problems and go through following the docs or I need first do something with the pom file?

**Comment by toskrip** [comment]
These 3 lines that I mentioned you can ignore. But "could not find artifact error" you need to solve. Because you need to be able to fetch ojdbc14-10g.jar via maven.

**Comment by shanzhu2021** [comment]
When I further check the pom files in the four dir, I found there are still problems as shown in the snapshots.
Which I have to fix and which I can run without change?

In addition, for the PostgreSQL database, I changed the port to 5434, where I should make change in the project?

<img width="739" alt="errors1" src="https://user-images.githubusercontent.com/77312250/105627574-adcb5a00-5e72-11eb-94d8-21d16de8385c.png">

**Comment by toskrip** [comment]
> LibreClinica-odm

jaxb2-maven-plugin does not work anymore in newest version of Java. We do development on java 8. There is a ticket open related to this issue (to allow development on newer java #173) but it is not resolved yet.

> the rest of missing dependencies

as mentioned above, those are dependencies only available from OC server. Obviously your intellij/maven configuration does not allow maven to fetch them. Alternatively, if you download them directly, you should be able to manually add them. Internet search for "manually install maven dependency" should give you an idea how to do it.

> database port

this can be done by changing datainfo.properties file after first deployment (it is described in docs Project in IDE > Debug section)

**Comment by shanzhu2021** [comment]
For the dependencies, I installed the local ojdbc manually, but I cannot find the two mockrunner files, so can I ignore these two and all the others? Otherwise, can you send me all the links if I need download and install locally.

For the database port, I find six datainfo.properties files in the project, but  from docs Project in IDE > Debug section, I do not need to change anything in these files, am I right?

**Comment by toskrip** [comment]
For manually downloading dependencies, you can browse them via this URL:
https://openclinica.jfrog.io/artifactory/libs-release/

OC still keeps this artifact repository open for everybody. But make sure that you install them properly, because from your screenshot it looks like the odbc dependency is still not recognised (missing).

Regarding datainfo.properties. Normally you do not want to modify the source files. When you compile and pack the LibreClinica into LibreClinica.war archive, the default version of datainfo.properties will be packed as well and after deployment in tomcat webapps folder you just copy that file (LibreClinica/WEB-INF/Classes/datainfo.properties) into folder that you create in TOMCAT_HOME/libreclinica.config/

And this is the file where you modify the configuration to pass your DB environment. This way we can keep config files that are in sources clean from any specific deployment environment details.

**Comment by shanzhu2021** [comment]
I installed ojdbc this way instead of using 'mavel install', it should be correct now?

<img width="863" alt="ojdbc14-installed" src="https://user-images.githubusercontent.com/77312250/105791824-977fe400-5fc1-11eb-8ca8-10a8a7812dcb.PNG">

In the same way, I installed mockrunners from your link, but have to change the scope from 'test' to 'system', otherwise, it reports errors, it now looks like:

<img width="870" alt="mockrunner-installed" src="https://user-images.githubusercontent.com/77312250/105810417-dd4ea380-5fe5-11eb-9922-b96720499da3.PNG">

is this the right way to do? Or I should use 'maven install'? Please advise me.

And all other errors in pom file of project and odm dirs can be ignored? Or What I need to do?

<img width="870" alt="pom-odm-errors" src="https://user-images.githubusercontent.com/77312250/105793753-a7e58e00-5fc4-11eb-8fb3-ab23015382f0.PNG">

**Comment by shanzhu2021** [comment]
I follow the docs and when I first execute Run > Debug, it reports an error before deploy it to tomcat, the screenshot is like:

<img width="861" alt="cdisc-errors" src="https://user-images.githubusercontent.com/77312250/105953992-0ab35400-60af-11eb-8dc2-4ef0d0451fe8.PNG">

What is this problem and where is the org.cdisc.ns.odm.v130_api?
Is this system can only built with Java 8?

**Reply by toskrip** [reply]
The application runs on newer Java, but for development/building currently Java 8 is required see ticket #173.

**Reply by shanzhu2021** [reply]
So the problems I met today is caused by the version of Java on my computer, and I have to lower my Java？

**Reply by toskrip** [reply]
Yes, to be concrete it is related to odm module. This module contains xsd schema that defines DTO classes related to marshalling/unmarshalling of CDISC data (with vendor specific extensions). We decided not to version control these classes (as they can be generated from the xsd schema during compilation time). This is exactly what the jaxb2-maven-plugin in odm module pom is doing and the result packed as LibreClinica-odm.jar is feeded as dependency for other modules in LibreClinica project. Unfortunately this maven plugin does not work anymore in newer java (only java 8) so we need to find different solution how to accomplish the same in order to make compilation and build on newer jdk possible (as described in the ticket).

**Comment by shanzhu2021** [comment]
I installed Java 8 within the IDEA, but the problem still exists:

<img width="1249" alt="java8" src="https://user-images.githubusercontent.com/77312250/106120892-e67c7380-6191-11eb-9084-dad2f14f4538.PNG">

and I also try jaxb2 v1.5 and v 2.2, the same problem:

<img width="507" alt="jaxb2_errors" src="https://user-images.githubusercontent.com/77312250/106121062-16c41200-6192-11eb-9b8f-dda8ab2fc2af.PNG">

There are still problem with cdisc, I can not find org.cdisc.ns.odm.v130_api.jar but only sorce code (still have problem with jaxb2), is there a link for the JAR?
<img width="876" alt="odmc_errors" src="https://user-images.githubusercontent.com/77312250/106121374-76bab880-6192-11eb-969a-a769f30d24c3.PNG">

**Reply by chae-reliatec** [reply]
The issue with antlr can be ignored. I never used IntelliJ IDEA but as far as I can tell it only says that the sources and the javadocs are missing for the package. These are not necessary to build LibreClinica.

**Reply by shanzhu2021** [reply]
So, how to run LibreClinica?

**Reply by chae-reliatec** [reply]
To run LibreClinica you can download the war archive from the LibreClinica webpage (https://www.libreclinica.org/download.html). But if I understand the topic correctly you try to setup a development environment to make changes to the source code. Are you stuck to IntellJ IDEA? I successfully set up a development environment in eclipse.

**Reply by chae-reliatec** [reply]
Tomas is the expert for the jaxb/CDISC/ODM topic but I believe that there is no org.cdisc.ns.odm.v130_api.jar generated. This all is packed into to LibreClinica-odm-2.2.jar. So you should concentrate on the jaxb problem since this seems to be the root cause of your issues. What happens when you just build the LibreClinica-ODM project?

**Comment by chae-reliatec** [comment]


**Reply by shanzhu2021** [reply]
Before I can build the LibreClinica-ODM project, the pom.xml has errors with jaxb2, so I ask what is the problem?
<img width="507" alt="jaxb2_errors" src="https://user-images.githubusercontent.com/77312250/106126761-c308f700-6198-11eb-8a3a-2f27889d8863.PNG">

**Reply by chae-reliatec** [reply]
Are there some error messages you can provide? I only see some red xml elements (schemaDirectory, bindingDirectory and bindingFiles).

**Comment by shanzhu2021** [comment]
The red xml elements reads "Element schemaDirectory is not allowed here".

**Reply by chae-reliatec** [reply]
This sounds like an error in IntellJ IDEA. Maybe it can not access the XML schema definition correctly but I do not think this is a real issue.
Are there other issues?

**Reply by shanzhu2021** [reply]
Others like "org.cdisc.ns.odm.v130_api does not exist"

**Reply by chae-reliatec** [reply]
Can you execute maven from the commandline?
 
`mvn clean package -pl odm`

**Reply by shanzhu2021** [reply]
yes, I will try.

**Reply by shanzhu2021** [reply]
It looks like work this time when I do mvn clean and then debug, thanks very much!
But finally it reports:
[2021-01-28 07:18:09,924] Artifact LibreClinica-web:war: Error during artifact deployment. See server log for details.
I will check it further!

**Comment by shanzhu2021** [comment]
After change to tomcat 8, I run the project successfully. Thanks all of your help!

**Comment by shanzhu2021** [comment]
Does the LibreClinica support multi-lingual? How to change it?

**Reply by toskrip** [reply]
For new topic/question next time please start new discussion. LibreClinica is using standard way to achieve internationalization for java application. The language specific terms are stored in properties files in:

resources/org/akaza/openclinica/i18n/

Not sure if it is possible to run multiple languages side by side... but some people did change those properties to achieve translated interface. Supported is a bit strong word, as we do not test LibreClinica functionality under translated conditions. Things can go wrong with the way how dates and number are processed for different locale settings. The original parent project from which LibreClinica is forked had in the past community afford to establish translated versions of properties files:

https://www.transifex.com/ccollins/openclinica

**Reply by shanzhu2021** [reply]
sure, thanks!
